sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/variant/variantExpressions.scala

```diff
@@ -114,6 +114,85 @@ case class IsVariantNull(child: Expression) extends UnaryExpression
     copy(child = newChild)
 }
 
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Convert a nested input (array/map/struct) into a variant where maps and structs are converted to variant objects which are unordered unlike SQL structs. Input maps can only have string keys.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(named_struct('a', 1, 'b', 2));
+       {"a":1,"b":2}
+      > SELECT _FUNC_(array(1, 2, 3));
+       [1, 2, 3]
+      > SELECT _FUNC_(array(named_struct('a', 1)));
+       [{"a":1}]
+      > SELECT _FUNC_(array(map("a", 2)));
+       [{"a":2}]
+  """,
+  since = "4.0.0",
+  group = "variant_funcs")
+// scalastyle:on line.size.limit
+case class ToVariantObject(child: Expression)
+    extends UnaryExpression
+    with NullIntolerant
+    with QueryErrorsBase {
+
+  override val dataType: DataType = VariantType
+
+  // Only accept nested types at the root but any types can be nested inside.
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val checkResult: Boolean = child.dataType match {
+      case ArrayType(elementType, _) =>
+        VariantGet.checkDataType(elementType, allowStructsAndMaps = true)
+      case MapType(_: StringType, valueType, _) =>
+        VariantGet.checkDataType(valueType, allowStructsAndMaps = true)
+      case StructType(fields) =>
+        fields.forall(f => VariantGet.checkDataType(f.dataType, allowStructsAndMaps = true))
```

Comment:
I feel like this could be simplified like:

```suggestion
      case _: StructType | _: ArrayType | _: MapType =>
        VariantGet.checkDataType(child.dataType, allowStructsAndMaps = true)
```
since `VariantGet.checkDataType` wil handle the all the type checking.